File: q0606/q0606.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    string tree2str(TreeNode *root)
    {
        string res = "";
        dfsHelper(root, res);
        return res;
    }

private:
    void dfsHelper(TreeNode *node, string &res)
    {

        // base case 1: node is null
        if (node == nullptr)
            return;

        res += to_string(node->val);

        // base case 2: node is leaf node
        if (node->left == nullptr && node->right == nullptr)
        {
            return;
        }

        // recurse on the left
        res += "(";
        dfsHelper(node->left, res);
        res += ")";

        // recurse on the right
        if (node->right != nullptr)
        {
            res += "(";
            dfsHelper(node->right, res);
            res += ")";
        }
    }
};

File: q0606/q0606.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    public String tree2str(TreeNode root) {
        StringBuilder sb = new StringBuilder();
        dfs(root, sb);
        return sb.toString();
    }

    private void dfs(TreeNode node, StringBuilder currString) {
        // base case 1
        if (node == null)
            return;

        currString.append(node.val);

        // base case 2
        if (node.left == null && node.right == null)
            return;

        // recurse on the left
        currString.append("(");
        dfs(node.left, currString);
        currString.append(")");

        // recurse on the right
        if (node.right != null) {
            currString.append("(");
            dfs(node.right, currString);
            currString.append(")");
        }
    }
}